### PR TITLE
Rework reconnect logic

### PR DESCRIPTION
**Pull request by mosquito**

### CODE REVIEW

File: tests/test_amqp.py [hunk 1]
```diff
@@ -1597,7 +1601,7 @@ async def test_connection_close(
         try:
             with pytest.raises(aio_pika.exceptions.ChannelPreconditionFailed):
                 msg = Message(bytes(shortuuid.uuid(), "utf-8"))
-                msg.delivery_mode = 8
+                msg.delivery_mode = 8       # type: ignore
```
Comment (Alviner) [inline tests/test_amqp.py:1604]: what's happened here? why mypy is struggling?

**Review by Alviner — APPROVED** [review]

**Review by Alviner — DISMISSED** [review]

File: tests/test_amqp.py (hunk 1 above)
Comment (mosquito) [reply tests/test_amqp.py:1604]: Because it's IntEnum

**Review by Alviner — APPROVED** [review]

**Review by Alviner — DISMISSED** [review]

File: aio_pika/abc.py [hunk 2]
```diff
@@ -114,20 +115,36 @@ async def __aexit__(
...
-    HeadersValue,
-    Set[HeadersValue],
-    Tuple[HeadersValue, ...],
-    FrozenSet[HeadersValue],
-]
-HeadersType = MutableMapping[str, HeadersPythonValues]
+HeadersType = Dict[str, FieldValue]
+
+
+class MessageInfo(TypedDict, total=False):
```
Comment (decaz) [inline aio_pika/abc.py:121]: Why not dataclass?

File: aio_pika/abc.py [hunk 3]
```diff
@@ -257,7 +265,17 @@ def __init__(
         arguments: Arguments,
         passive: bool = False,
     ):
-        raise NotImplementedError
+        raise NotImplementedError(
+            dict(
```
Comment (decaz) [inline aio_pika/abc.py:269]: Why do we need this data here? It's the only place where `NotImplementedError` is being raised with data..

File: aio_pika/abc.py (hunk 2 above)
Comment (mosquito) [reply aio_pika/abc.py:121]: dataclass with total=False is a very strange solution. Message info it's a way for print message to logs etc. The dictionary is a completely relevant for this.

File: aio_pika/abc.py (hunk 3 above)
Comment (mosquito) [reply aio_pika/abc.py:269]: This code is a hack for linters it's never being called in the runtime

File: pyproject.toml [hunk 4]
```diff
@@ -0,0 +1,107 @@
+[tool.poetry]
+name = "aio-pika"
+version = "9.0.0"
+description = "Wrapper for the aiormq for asyncio and humans"
```
Comment (decaz) [inline pyproject.toml:4]: Wrapper **over** aiormq ... would be better, I guess.

File: aio_pika/robust_connection.py [hunk 5]
```diff
@@ -95,11 +97,64 @@ async def __connection_attempt(
...
+                await asyncio.sleep(self.reconnect_interval)
+            except Exception:
+                log.exception(
+                    "Reconnect attempt failed %s. "
+                    "Retrying after %r seconds.",
+                    self, self.reconnect_interval,
+                )
+                await asyncio.sleep(self.reconnect_interval)
+
+            await asyncio.sleep(self.reconnect_interval)
```
Comment (decaz) [inline aio_pika/robust_connection.py:153]: I guess two sleeps above are redundant, are they?

Comment (mosquito) [reply aio_pika/robust_connection.py:153]: Nope previous one in the except block

File: pyproject.toml (hunk 4 above)
Comment (mosquito) [reply pyproject.toml:4]: around?

Comment (decaz) [reply pyproject.toml:4]: Yes, it's suitable 👍🏻

File: aio_pika/robust_connection.py (hunk 5 above)
Comment (decaz) [reply aio_pika/robust_connection.py:153]: Yes, but this sleep will run after these `except` blocks.

Comment (mosquito) [reply aio_pika/robust_connection.py:151]: You right

**Review by decaz — APPROVED** [review]
@mosquito great work, thank you!

**Pull request merged**